src/transforms.rs

```diff
@@ -0,0 +1,254 @@
+    ExpressionTransform(Expression<'static>),
+}
+
+pub(crate) struct TransformContext {
+    transforms: HashMap<String, MessageTransform>,
+}
+
+impl TransformContext {
+    pub(crate) fn new(transforms: HashMap<String, MessageTransform>) -> Self {
+        Self { transforms }
+    }
+
+    pub(crate) fn transform<M>(
+        &self,
+        value: &mut Value,
+        kafka_message: &M,
+    ) -> Result<(), TransformError>
+    where
+        M: Message,
+    {
+        let data = jmespath::Variable::from(value.clone());
+
+        match value.as_object_mut() {
+            Some(map) => {
+                for (property, message_transform) in self.transforms.iter() {
+                    let property_value = match message_transform {
+                        MessageTransform::ExpressionTransform(expression) => {
+                            let variable = expression.search(data.clone())?;
+                            serde_json::to_value(variable)?
+                        }
+                        MessageTransform::KafkaMetaTransform(meta_property) => {
+                            match meta_property {
+                                KafkaMetaProperty::Partition => {
+                                    serde_json::to_value(kafka_message.partition())?
+                                }
+                                KafkaMetaProperty::Offset => {
+                                    serde_json::to_value(kafka_message.offset())?
+                                }
+                                KafkaMetaProperty::Topic => {
+                                    serde_json::to_value(kafka_message.topic())?
+                                }
+                                KafkaMetaProperty::Timestamp => {
+                                    serde_json::to_value(kafka_message.timestamp().to_millis())?
+                                }
+                            }
+                        }
+                    };
+
+                    map.insert(property.to_string(), property_value);
+                }
+                Ok(())
+            }
+            _ => Err(TransformError::ValueNotAnObject {
+                value: value.to_owned(),
+            }),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rdkafka::message::OwnedMessage;
+
+    use super::*;
+    use serde_json::json;
+
+    #[test]
+    fn transforms_with_substr() {
+        let mut test_value = json!({
+            "name": "A",
+            "modified": "2021-03-16T14:38:58Z",
+        });
+
+        let test_message = OwnedMessage::new(
+            Some(test_value.to_string().into_bytes()),
+            None,
+            "test".to_string(),
+            rdkafka::Timestamp::NotAvailable,
+            0,
+            0,
+            None,
+        );
+
+        let mut transforms = HashMap::new();
+
+        transforms.insert(
+            "modified_date".to_string(),
+            "substr(modified, '0', '10')".to_string(),
```

Comment:
I understand that this isn't a "real" function call, but why are the integer arguments in strings and then `modified` has no quotes? I'm having trouble parsing what syntax means here with this `substr` function

Reply:
So I finally figured out why passing numbers wasn't working for me... JMESPath requires backticks around literals.  The new syntax that actually accepts numbers where they belong is:

```
substr(modified, `0`, `10`)
```

I probably need to stuff a JMESPath syntax link somewhere in here...

Reply:
Added better documentation about the transforms to the usage command in https://github.com/delta-io/kafka-delta-ingest/pull/4/commits/d728dc0a6ea8d217cc9643de7f334fd7bc7324cc